pkg/ottl/functions.go

```diff
@@ -32,65 +55,67 @@ func (p *Parser[K]) newFunctionCall(inv invocation) (Expr[K], error) {
 	if !ok {
 		return Expr[K]{}, fmt.Errorf("undefined function %v", inv.Function)
 	}
-	args, err := p.buildArgs(inv, reflect.TypeOf(f))
+	args := f.CreateEmptyArguments()
+
+	// Pointer values are necessary to fulfill the Go reflection
+	// settability requirements. Non-pointer values are not
+	// modifiable through reflection.
+	if reflect.TypeOf(args).Kind() != reflect.Pointer {
+		return Expr[K]{}, fmt.Errorf("factory for %s must return a pointer to an Arguments value in its CreateEmptyArguments method", inv.Function)
+	}
+
+	err := p.buildArgs(inv, reflect.ValueOf(args).Elem())
 	if err != nil {
 		return Expr[K]{}, fmt.Errorf("error while parsing arguments for call to '%v': %w", inv.Function, err)
 	}
 
-	returnVals := reflect.ValueOf(f).Call(args)
-
-	if returnVals[1].IsNil() {
-		err = nil
-	} else {
-		err = returnVals[1].Interface().(error)
+	fn, err := f.CreateFunction(context.Background(), p.telemetrySettings, args)
+	if err != nil {
+		return Expr[K]{}, fmt.Errorf("couldn't create function")
 	}
 
-	return Expr[K]{exprFunc: returnVals[0].Interface().(ExprFunc[K])}, err
+	return Expr[K]{exprFunc: fn}, err
 }
 
-func (p *Parser[K]) buildArgs(inv invocation, fType reflect.Type) ([]reflect.Value, error) {
-	var args []reflect.Value
-	// Some function arguments may be intended to take values from the calling processor
-	// instead of being passed by the caller of the OTTL function, so we have to keep
-	// track of the index of the argument passed within the DSL.
-	// e.g. TelemetrySettings, which is provided by the processor to the OTTL Parser struct.
-	DSLArgumentIndex := 0
-	for i := 0; i < fType.NumIn(); i++ {
-		argType := fType.In(i)
+func (p *Parser[K]) buildArgs(inv invocation, argsVal reflect.Value) error {
+	if len(inv.Arguments) > argsVal.NumField() {
+		return fmt.Errorf("too many arguments")
+	}
+
+	argsType := argsVal.Type()
+
+	for i := 0; i < argsVal.NumField(); i++ {
+		field := argsVal.Field(i)
+		fieldType := field.Type()
+
+		fieldTag, ok := argsType.Field(i).Tag.Lookup("ottlarg")
 
-		arg, isInternalArg := p.buildInternalArg(argType)
-		if isInternalArg {
-			args = append(args, reflect.ValueOf(arg))
-			continue
+		if !ok {
+			return fmt.Errorf("no `arg` struct tag on Arguments field %s", argsType.Field(i).Name)
 		}
 
-		if DSLArgumentIndex >= len(inv.Arguments) {
-			return nil, fmt.Errorf("not enough arguments")
+		argNum, err := strconv.Atoi(fieldTag)
+
+		if err != nil {
+			return fmt.Errorf("invalid field tag argument number :%w", err)
 		}
 
-		argVal := inv.Arguments[DSLArgumentIndex]
+		argVal := inv.Arguments[argNum]
```

Comment:
Since we haven't checked that `len(inv.Arguments) == argsVal.NumField()` and only checked that it is not greater than, I think there is an opportunity for index-out-of-bounds here.

Reply:
I just noticed that we don't do any bounds checking whatsoever on this, so I've enforced the bounds on the value.